src/intrinsics/ecma262/Object.js

```diff
@@ -61,6 +61,15 @@ export default function(realm: Realm): NativeFunctionValue {
     let to = To.ToObjectPartial(realm, target);
     let to_must_be_partial = false;
 
+    if (to.isSimpleObject()) {
+      // isSimpleObject() is lazily calculated based on properties for normal objects.
+      // However we might later make it partial in this function. In that case,
+      // the default will be that the object is *not* simple. But we know that
+      // it was simple before, and that Object.assign() can't possibly copy any
+      // getters or setters onto it. Therefore, it is safe to claim it stays simple.
+      to.makeSimple();
```

Comment:
This makes this object permanently simple even if we do:

```js
var x = Object.assign({}, partial);
var y = 0;
Object.defineProperty(x, 'foo', { get() { y++ } });
Object.assign({}, x);
result = y;
```

We need that defineProperty to unmark this as simple.

Reply:
That's what I thought but I asked @hermanventer and he said

>The idea of a simple object is simply that it's safe to read and write unknown properties. If there is a getter of setter involved that is OK as long as there is no observable side effects. **Of course if the property is known and we have the definition of the getter or setter then that is OK too.**

and

>Demoting a simple object is not a supported scenario. It is probably not a good idea.

So I'm confused over this. I'm also not sure what assumptions Prepack codebase makes around "simple" (i.e. does it treat it as "has no getters/setters" or "has no unknown getters/setters") in practice, regardless of the original intention.

For example would this be safe if there were known getters/setters:

https://github.com/facebook/prepack/blob/6203b6b125e916815fac848d06ca86ed7f73d4dc/src/evaluators/BinaryExpression.js#L133

?

This

https://github.com/facebook/prepack/blob/6203b6b125e916815fac848d06ca86ed7f73d4dc/src/methods/to.js#L783

?

I'm not sure. The fact that it's used in so few places also makes it hard for me to tell what it's supposed to represent in practice, and if it's consistent today or not.

Reply:
If that's the case, then it seems to me that we should only set this explicit flag if we end up in the unknown partial case.